eth/downloader/downloader.go

```diff
@@ -857,8 +880,18 @@ func (d *Downloader) fetchHeaders(p *peerConnection, from uint64, pivot uint64)
 					return errCancelHeaderFetch
 				}
 				from += uint64(len(headers))
+				getHeaders(from)
+			} else {
+				// No headers delivered, or all of them being delayed, sleep a bit and retry
+				p.log.Trace("All headers delayed, waiting")
+				select {
```

Comment:
I don't know the full details about how this part works -- but you introduce a sleep here (`3s`) which wasn't there previously. Are you sure that's 1. sufficient time and 2. doesn't block other important things, like downloader events that cannot be delivered?

Reply:
This mirrors line 826, which did the same thing, just when we fully exhausted our headers.